src/sync/rpc.ts

```diff
@@ -0,0 +1,277 @@
+/**
+ * @module sync
+ */
+
+import assert from "assert";
+import BN from "bn.js";
+import {hashTreeRoot} from "@chainsafe/ssz";
+import {
+  bytes32, Slot, number64,
+  BeaconBlockHeader, BeaconBlockBody,
+  RequestBody, Hello, Goodbye, Status,
+  BeaconBlockRootsRequest, BeaconBlockRootsResponse,
+  BeaconBlockHeadersRequest, BeaconBlockHeadersResponse,
+  BeaconBlockBodiesRequest, BeaconBlockBodiesResponse,
+  BeaconStatesRequest, BeaconStatesResponse,
+} from "../types";
+import {intDiv} from "../util/math";
+import {IBeaconDb} from "../db";
+import {BeaconChain} from "../chain";
+import {INetwork, IPeer} from "../network";
+import {Method, RequestId} from "../network/codec";
+import {getEmptyBlockBody} from "../chain/genesis";
+import {ZERO_HASH} from "../constants";
+
+/**
+ * The SyncRpc module handles app-level requests / responses from other peers,
+ * fetching state from the chain and database as needed.
+ */
+export class SyncRpc {
+  private db: IBeaconDb;
+  private chain: BeaconChain;
+  private network: INetwork;
+  public constructor(opts, {db, chain, network}) {
+    this.db = db;
+    this.chain = chain;
+    this.network = network;
+  }
+
+  // create common requests
+
+  public async createHello(): Promise<Hello> {
+    let bestSlot, bestRoot, latestFinalizedEpoch, latestFinalizedRoot;
+    if (!this.chain.genesisTime) {
+      bestSlot = 0;
+      bestRoot = ZERO_HASH;
+      latestFinalizedEpoch = 0;
+      latestFinalizedRoot = ZERO_HASH;
+    } else {
+      bestSlot = await this.db.getChainHeadSlot();
+      bestRoot = await this.db.getBlockRoot(bestSlot);
+      const state = await this.db.getState();
+      latestFinalizedEpoch = state.finalizedEpoch;
+      latestFinalizedRoot = state.finalizedRoot;
+    }
+    return {
+      networkId: this.chain.networkId,
+      chainId: this.chain.chainId,
+      latestFinalizedRoot,
+      latestFinalizedEpoch,
+      bestRoot,
+      bestSlot,
+    };
+  }
+
+  public createStatus(): Status {
+    return {
+      sha: Buffer.alloc(32),
+      userAgent: Buffer.from("Lodestar"),
+      timestamp: intDiv(Date.now(), 1000),
+    };
+  }
+
+  // send outgoing requests
+
+  private async handshake(peerInfo: PeerInfo): Promise<void> {
+    const randomDelay = Math.floor(Math.random() * 5000);
+    await new Promise((resolve) => setTimeout(resolve, randomDelay));
```

Comment:
i think in some test utils there is sleep method

Reply:
That said what does this delay do?

Reply:
It gives one party a chance to initialize a handshake, instead of both trying at once.